### PR TITLE
[SPARK-27590][CORE] do not consider skipped tasks when scheduling speculative tasks

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/24375

When `TaskSetManager` skips a task because its corresponding partition is already completed by other `TaskSetManager`s, we should not consider the duration of the task that is finished by other `TaskSetManager`s to schedule the speculative tasks of this `TaskSetManager`.

## How was this patch tested?

updated test case

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @squito @jiangxb1987

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 1]
```diff
@@ -1038,7 +1035,11 @@ private[spark] class TaskSetManager(
     val minFinishedForSpeculation = (speculationQuantile * numTasks).floor.toInt
     logDebug("Checking for speculative tasks: minFinished = " + minFinishedForSpeculation)
 
-    if (tasksSuccessful >= minFinishedForSpeculation && tasksSuccessful > 0) {
+    // It's possible that a task is marked as completed by the scheduler, then the size of
+    // `successfulTaskDurations` may not equal to `tasksSuccessful`. Here we should only count the
+    // tasks that are submitted by this `TaskSetManager` and are completed successfully.
+    val numSuccessfulTasks = successfulTaskDurations.size()
+    if (numSuccessfulTasks >= minFinishedForSpeculation && numSuccessfulTasks > 0) {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: unrelated to your change, but a potential cleanup here -- why not compute `minFinishedForSpeculation` in the construction, and even set it to `math.max(1, (speculationQuantile * numTasks).floor.toInt)`?  Then you also don't need the extra `numSuccessfulTasks > 0` here.

**Review by squito — COMMENTED** [review]
lgtm, just a small suggestion on a cleanup

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 1 above)
Comment (jiangxb1987) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: +1

**Comment by jiangxb1987** [comment]
LGTM

**Comment by cloud-fan** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 2]
```diff
@@ -1032,14 +1032,14 @@ private[spark] class TaskSetManager(
       return false
     }
     var foundTasks = false
-    val minFinishedForSpeculation = (speculationQuantile * numTasks).floor.toInt
+    val minFinishedForSpeculation = math.max((speculationQuantile * numTasks).floor.toInt, 1)
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: sorry I had actually meant to just compute this once in the constructor, eg. around line 70

**Comment by squito** [comment]
merged to master.  Thanks @cloud-fan !

**Comment by cloud-fan** [comment]
@squito thanks for the review and merge!

**Pull request closed without merging**